### PR TITLE
Update Rust and use the newly-stable std::ptr::NonNull

### PR DESCRIPTION
**Pull request by SimonSapin**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19829)

### CODE REVIEW

File: components/script/dom/bindings/iterable.rs [hunk 1]
```diff
@@ -106,7 +106,7 @@ impl<T: DomObject + JSTraceable + Iterable> IterableIterator<T> {
         self.index.set(index + 1);
         result.map(|_| {
             assert!(!rval.is_null());
-            unsafe { NonNullJSObjectPtr::new_unchecked(rval.get()) }
+            unsafe { NonNull::new_unchecked(rval.get()) }
```
Comment (nox) [inline components/script/dom/bindings/iterable.rs]: Use `new` and `expect`.

File: components/script/dom/imagedata.rs [hunk 2]
```diff
@@ -159,8 +159,8 @@ impl ImageDataMethods for ImageData {
 
     #[allow(unsafe_code)]
     // https://html.spec.whatwg.org/multipage/#dom-imagedata-data
-    unsafe fn Data(&self, _: *mut JSContext) -> NonNullJSObjectPtr {
+    unsafe fn Data(&self, _: *mut JSContext) -> NonNull<JSObject> {
         assert!(!self.data.get().is_null());
-        NonNullJSObjectPtr::new_unchecked(self.data.get())
+        NonNull::new_unchecked(self.data.get())
```
Comment (nox) [inline components/script/dom/imagedata.rs]: Use `new` and `expect`.

File: components/script/dom/testbinding.rs [hunk 3]
```diff
@@ -152,20 +152,20 @@ impl TestBindingMethods for TestBinding {
     }
     fn SetUnion9Attribute(&self, _: ByteStringOrLong) {}
     #[allow(unsafe_code)]
-    unsafe fn ArrayAttribute(&self, cx: *mut JSContext) -> NonNullJSObjectPtr {
+    unsafe fn ArrayAttribute(&self, cx: *mut JSContext) -> NonNull<JSObject> {
         rooted!(in(cx) let array = JS_NewUint8ClampedArray(cx, 16));
         assert!(!array.is_null());
-        NonNullJSObjectPtr::new_unchecked(array.get())
+        NonNull::new_unchecked(array.get())
```
Comment (nox) [inline components/script/dom/testbinding.rs]: Use `new` and `expect`.

File: components/script/dom/xmlhttprequest.rs [hunk 4]
```diff
@@ -1120,19 +1120,19 @@ impl XMLHttpRequest {
 
     // https://xhr.spec.whatwg.org/#arraybuffer-response
     #[allow(unsafe_code)]
-    unsafe fn arraybuffer_response(&self, cx: *mut JSContext) -> Option<NonNullJSObjectPtr> {
+    unsafe fn arraybuffer_response(&self, cx: *mut JSContext) -> Option<NonNull<JSObject>> {
         // Step 1
         let created = self.response_arraybuffer.get();
         if !created.is_null() {
-            return Some(NonNullJSObjectPtr::new_unchecked(created));
+            return Some(NonNull::new_unchecked(created));
```
Comment (nox) [inline components/script/dom/xmlhttprequest.rs]: Use `new`.

**Review by nox — APPROVED** [review]
Looks good to me, feel free to handle my remarks as a follow-up issue. Would be nice if you filed an issue about auditing all the `new_unchecked` calls in general, some of them scare me.

**Comment by CYBAI** [comment]
```rust
   Compiling libservo v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/components/servo)
   Compiling glutin_app v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/ports/glutin)
   Compiling embedding v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/ports/cef)
error: unnecessary parentheses around function argument
  --> ports/cef/eutil.rs:33:34
   |
33 |     let object = libc::calloc(1, (mem::size_of::<Base>() + mem::size_of::<Extra>())) as
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
   |
   = note: `-D unused-parens` implied by `-D warnings`

error: unnecessary parentheses around function argument
   --> ports/cef/wrappers.rs:202:36
    |
202 |             let ptr = libc::malloc(((buffer.len() + 1) * 2)) as *mut c_ushort;
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses

error: aborting due to 2 previous errors

error: Could not compile `embedding`.
```

**Comment by jdm** [comment]
```
  ▶ TIMEOUT [expected PASS] /css/CSS2/box-display/containing-block-009.xht
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 17.3.0-devel
  │ 2018-01-22 10:30:57.066 servo[81885:285780639] Metadata.framework [Error]: couldn't get the client port
  │ Stack trace for thread "WRRenderBackend#0"
  │ stack backtrace:
  │    0:        0x10f8f3aee - backtrace::backtrace::trace::h1572cbde15e97645
  │    1:        0x10f8f42dc - backtrace::capture::Backtrace::new::h0ca27cd1e659fd9b
  │    2:        0x10d3219a0 - servo::install_crash_handler::handler::hef6d72ec40496654
  │    3:     0x7fff8f88ff19 - _sigtramp
  │    4:        0x10ef1a4c7 - _$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::contains_key::h6f7b7a7f90f60aa7
  │    5:        0x10eed3f15 - webrender::glyph_rasterizer::_$LT$impl$u20$webrender..platform..macos..font..FontContext$GT$::add_font::h855f3dfb504b8fe4
  │    6:        0x10ef64457 - webrender::glyph_rasterizer::GlyphRasterizer::add_font::h8f362be6e920367e
  │    7:        0x10eefea16 - webrender::resource_cache::ResourceCache::add_font_template::hd9fa88ab61ef0758
  │    8:        0x10eefd306 - webrender::resource_cache::ResourceCache::update_resources::h4a4d139987a02082
  │    9:        0x10ef4d9c9 - webrender::render_backend::RenderBackend::run::h3fba174366c9d443
  │   10:        0x10ee879e1 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3370e1c89a690f63
  │   11:        0x10eeb9b1d - std::panicking::try::do_call::h939a1adbd2b681c9 (.llvm.BA14D8D4)
  │   12:        0x10f923fae - __rust_maybe_catch_panic
  │   13:        0x10ef59f93 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::heff2fe36abcdc229
  │   14:        0x10f90cd67 - std::sys_common::thread::start_thread::h1ef8cf6b973abe77
  │   15:        0x10f9100e8 - std::sys::unix::thread::Thread::new::thread_start::h88d15e6243c319fe
  │   16:     0x7fff91ccf059 - _pthread_body
  │   17:     0x7fff91ccefd6 - _pthread_start
  │ Stack trace for thread "WRRenderBackend#0"
  │ stack backtrace:
  │    0:        0x10f8f3aee - backtrace::backtrace::trace::h1572cbde15e97645
  │    1:        0x10f8f42dc - backtrace::capture::Backtrace::new::h0ca27cd1e659fd9b
  │    2:        0x10d3219a0 - servo::install_crash_handler::handler::hef6d72ec40496654
  │    3:     0x7fff8f88ff19 - _sigtramp
  │    4:        0x10d321a14 - servo::install_crash_handler::handler::hef6d72ec40496654
  │    5:     0x7fff8f88ff19 - _sigtramp
  │    6:        0x10ef1a4c7 - _$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::contains_key::h6f7b7a7f90f60aa7
  │    7:        0x10eed3f15 - webrender::glyph_rasterizer::_$LT$impl$u20$webrender..platform..macos..font..FontContext$GT$::add_font::h855f3dfb504b8fe4
  │    8:        0x10ef64457 - webrender::glyph_rasterizer::GlyphRasterizer::add_font::h8f362be6e920367e
  │    9:        0x10eefea16 - webrender::resource_cache::ResourceCache::add_font_template::hd9fa88ab61ef0758
  │   10:        0x10eefd306 - webrender::resource_cache::ResourceCache::update_resources::h4a4d139987a02082
  │   11:        0x10ef4d9c9 - webrender::render_backend::RenderBackend::run::h3fba174366c9d443
  │   12:        0x10ee879e1 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3370e1c89a690f63
  │   13:        0x10eeb9b1d - std::panicking::try::do_call::h939a1adbd2b681c9 (.llvm.BA14D8D4)
  │   14:        0x10f923fae - __rust_maybe_catch_panic
  │   15:        0x10ef59f93 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::heff2fe36abcdc229
  │   16:        0x10f90cd67 - std::sys_common::thread::start_thread::h1ef8cf6b973abe77
  │   17:        0x10f9100e8 - std::sys::unix::thread::Thread::new::thread_start::h88d15e6243c319fe
  │   18:     0x7fff91ccf059 - _pthread_body
  └   19:     0x7fff91ccefd6 - _pthread_start
```
Terrifying!

**Pull request merged**